libraries/GCS_MAVLink/GCS_MAVFTP.cpp

```diff
@@ -0,0 +1,717 @@
+        errorCode = _ftp_Reset(payload, response_payload);
+        break;
+
+    case kCmdListDirectory:
+        errorCode = _ftp_List(payload, response_payload);
+        break;
+
+    case kCmdOpenFileRO:
+        errorCode = _ftp_Open(payload, response_payload, O_RDONLY);
+        break;
+
+    case kCmdCreateFile:
+        errorCode = _ftp_Open(payload, response_payload, O_CREAT | O_EXCL | O_WRONLY);
+        break;
+
+    case kCmdOpenFileWO:
+        errorCode = _ftp_Open(payload, response_payload, O_CREAT | O_WRONLY);
+        break;
+
+    case kCmdReadFile:
+        errorCode = _ftp_Read(payload, response_payload);
+        break;
+
+    case kCmdBurstReadFile:
+        response_payload->offset = 0;
+        response_payload->size = 0;
+        response_payload->burst_complete = false;
+        errorCode = _ftp_Burst(payload, target_system_id);
+        break;
+
+    case kCmdWriteFile:
+        errorCode = _ftp_Write(payload, response_payload);
+        break;
+
+    case kCmdRemoveFile:
+        errorCode = _ftp_RemoveFile(payload, response_payload);
+        break;
+
+    case kCmdRename:
+        errorCode = _ftp_Rename(payload, response_payload);
+        break;
+
+    case kCmdTruncateFile:
+        errorCode = _ftp_TruncateFile(payload, response_payload);
+        break;
+
+    case kCmdCreateDirectory:
+        errorCode = _ftp_CreateDirectory(payload, response_payload);
+        break;
+
+    case kCmdRemoveDirectory:
+        errorCode = _ftp_RemoveDirectory(payload, response_payload);
+        break;
+
+    case kCmdCalcFileCRC32:
+        errorCode = _ftp_CalcFileCRC32(payload, response_payload);
+        break;
+
+    default:
+        errorCode = kErrUnknownCommand;
+        break;
+    }
+
+out:
+    response_payload->seq_number = payload->seq_number + 1;
+
+    // handle success vs. error
+    if (errorCode == kErrNone) {
+        response_payload->req_opcode = payload->opcode;
+        response_payload->opcode = kRspAck;
+    }
+    else {
+        response_payload->req_opcode = payload->opcode;
+        response_payload->opcode = kRspNak;
+        response_payload->size = 1;
+        response_payload->data[0] = errorCode;
+        if (errorCode == kErrFailErrno) {
+            response_payload->size = 2;
+            response_payload->data[1] = _ftp_errno;
+        }
+    }
+
+    //Send Response
+    _ftp_response.target_system = target_system_id;
+    mavlink_msg_file_transfer_protocol_send_struct(chan, &_ftp_response);
+}
+
+GCS_MAVLINK::FTPErrorCode GCS_MAVLINK::_ftp_List(FTPPayloadHeader *request_payload, FTPPayloadHeader *response_payload)
+{
+    FTPErrorCode errorCode = kErrNone;
+    unsigned offset = 0;
+    char* dir_name;
+    _set_current_abs_path(_data_as_cstring(request_payload));
+    dir_name = _ftp_current_abs_path;
+    DIR *dp = opendir(dir_name);
+
+    if (dp == nullptr) {
+        gcs().send_statustext(MAV_SEVERITY_CRITICAL, (1<<chan), "FTP: can't open path (file system corrupted?)");
```

Comment:
this is going to blat onto the MP console, and overwrite. Is it really needed? Are error codes not enough?